File: src/main/java/org/radarcns/util/Comparison.java
package org.radarcns.util;

import java.util.function.BiFunction;
import java.util.function.Function;

public interface Comparison<V> extends BiFunction<V, V, Integer> {
    static <V, T extends Comparable<T>> Comparison<V> compare(Function<V, T> property) {
        return (a, b) -> property.apply(a).compareTo(property.apply(b));
    }

    default <T extends Comparable<T>> Comparison<V> then(Function<V, T> property) {
        return (a, b) -> {
            int ret = apply(a, b);
            if (ret != 0) {
                return ret;
            }
            return compare(property).apply(a, b);
        };
    }
}

File: src/main/kotlin/org/radarbase/util/Comparison.kt
package org.radarbase.util

fun interface Comparison<V> : (V, V) -> Int {
    companion object {
        fun <V, T : Comparable<T>> compare(property: (V) -> T): Comparison<V> =
            Comparison { a, b -> property(a).compareTo(property(b)) }
    }

    fun <T : Comparable<T>> then(property: (V) -> T): Comparison<V> = Comparison { a, b ->
        val ret = this(a, b)
        if (ret != 0) ret else compare(property)(a, b)
    }
}
